sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultStringTypes.scala

```diff
@@ -0,0 +1,216 @@
+    }
+  }
+
+  /**
+   * Transforms the given plan, by transforming all expressions in its operators to use the given
+   * new type instead of the default string type.
+   */
+  private def transformPlan(plan: LogicalPlan, newType: StringType): LogicalPlan = {
+    plan resolveOperators { operator =>
+      operator resolveExpressionsUp { expression =>
+        transformExpression.applyOrElse((expression, newType), fallbackToExpression)
+      }
+    }
+  }
+
+  private def fallbackToExpression(tuple: (Expression, StringType)): Expression = tuple._1
+
+  /**
+   * Transforms the given expression, by changing all default string types to the given new type.
+   */
+  private def transformExpression: PartialFunction[(Expression, StringType), Expression] = {
+    case (columnDef: ColumnDefinition, newType) if hasDefaultStringType(columnDef.dataType) =>
+      columnDef.copy(dataType = replaceDefaultStringType(columnDef.dataType, newType))
+
+    case (cast: Cast, newType) if hasDefaultStringType(cast.dataType) =>
+      cast.copy(dataType = replaceDefaultStringType(cast.dataType, newType))
+
+    case (Literal(value, dt), newType) if hasDefaultStringType(dt) =>
+      Literal(value, replaceDefaultStringType(dt, newType))
+  }
+
+  private def hasDefaultStringType(dataType: DataType): Boolean =
+    dataType.existsRecursively(isDefaultStringType)
+
+  private def isDefaultStringType(dataType: DataType): Boolean = {
+    dataType match {
+      case _: TemporaryStringType =>
+        !replaceWithTempType
+      case st: StringType =>
+        // should only return true for StringType object and not StringType("UTF8_BINARY")
+        st.eq(StringType)
+      case _ => false
+    }
+  }
+
+  private def replaceDefaultStringType(dataType: DataType, newType: StringType): DataType = {
+    dataType.transformRecursively {
+      case currentType: StringType if isDefaultStringType(currentType) =>
+        if (replaceWithTempType && currentType == newType) {
+          getTemporaryStringType(currentType)
+        } else {
+          newType
+        }
+    }
+  }
+
+  private def getTemporaryStringType(forType: StringType): StringType = {
+    if (forType == StringType) {
+      TemporaryStringType(StringType("UTF8_LCASE").collationId)
+    } else {
+      TemporaryStringType(StringType.collationId)
+    }
+  }
+
+  private def replaceColumnTypes(
+      colTypes: Seq[QualifiedColType],
+      newType: StringType): Seq[QualifiedColType] = {
+    colTypes.map {
+      case colWithDefault if hasDefaultStringType(colWithDefault.dataType) =>
+        val replaced = replaceDefaultStringType(colWithDefault.dataType, newType)
+        colWithDefault.copy(dataType = replaced)
+
+      case col => col
+    }
+  }
+}
+
+case object ResolveDefaultStringTypes
+  extends ResolveDefaultStringTypes(replaceWithTempType = true) {}
+
+case object ResolveDefaultStringTypesWithoutTempType
```

Comment:
I don't like this fake rule approach, how about this
```
object ResolveDefaultStringTypes {
  def apply.. {
    val newPlan = apply0...
    if (plan.ne(newPlan)) {
      apply0...
    }
  }

  def apply0...
}
```

Reply:
The issue here is how do we know in the transformation if we need to replace with temporary type or not? We need to have some  state which the case objects  can provide

Reply:
we don't need a state (the `replaceWithTempType` flag) because `apply0` can blindly do
```
case StringType => 
  if (newType == StringType) TemporaryStringType else newType
case TemporaryStringType =>
  StringType(0)
```

The assumption is that `object StringType` and `TemporaryStringType` won't co-exist. We always call `apply0` once first, if `!newPlan.fastEquals(plan)`, it means replacement happened (though `TemporaryStringType` may not be used), so we call `apply0` again.

Reply:
Thanks for pointing out the error in resolve, now we can make it much simpler like you mentioned above!